### ISSUE TITLE
v0.13.2 Crashes vscode extension host on Mac

### ISSUE THREAD

**Issue by cameronvoell** [foambubble/foam#620]
- Foam version: v0.13.2 
- Platform: Mac (Big Sur 11.3.1, vscode Version: 1.56.0)
- Issue occur on the [foam template](https://github.com/foambubble/foam-template) repo: No

**Summary**

Update to 0.13.2 Foam extension crashes vscode Extension host on Mac
Reverting to 0.13.1 fixes issue. 

**Steps to reproduce**
1. Install v0.13.2
2. Open foam directory and try any foam command
3. vscode error in bottom right: Extension host terminated unexpectedly

**Additional information**

Log from debug console in vscode:
<details>

```
electron/js2c/renderer_init.js:61 The vm module of Node.js is deprecated in the renderer process and will be removed.
t._load @ electron/js2c/renderer_init.js:61
console.ts:137 [Extension Host] (node:52545) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.(Use `Code Helper (Renderer) --trace-deprecation ...` to show where the warning was created)
S @ console.ts:137
event.ts:487 [g6h] potential listener LEAK detected, having 175 listeners already. MOST frequent listener (1):
check @ event.ts:487
event.ts:488     at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:58:58523
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:58:61006
    at new Promise (<anonymous>)
    at Object.N [as toPromise] (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:58:60987)
    at _.whenInitialized (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:76:4475)
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:76:2720
    at w (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:68:11210)
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:76:2710
    at new Promise (<anonymous>)
    at _.requestPromise (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:76:2252)
    at Object.call (file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:76:1941)
    at file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:76:7092
check @ event.ts:488
localProcessExtensionHost.ts:272 Extension Host
localProcessExtensionHost.ts:273 Debugger attached.
log.ts:290  WARN UNRESPONSIVE extension host, 'foam.foam-vscode' took 71% of 5323.146ms, saved PROFILE here: 'file:///var/folders/vq/71_sc5r53h55bltlxmn9mxyc0000gn/T/exthost-eea25b.cpuprofile' Array(2)
localProcessExtensionHost.ts:272 Extension Host
localProcessExtensionHost.ts:273 <--- Last few GCs --->[52545:0x2600000000]    78813 ms: Scavenge (reduce) 3832.5 (3949.0) -> 3831.7 (3949.5) MB, 2.4 / 0.0 ms  (average mu = 0.591, current mu = 0.467) allocation failure [52545:0x2600000000]    79950 ms: Scavenge (reduce) 3832.8 (3949.5) -> 3832.0 (3949.5) MB, 1132.8 / 0.0 ms  (average mu = 0.591, current mu = 0.467) allocation failure [52545:0x2600000000]    79959 ms: Scavenge (reduce) 3833.0 (3949.5) -> 3832.2 (3949.5) MB, 4.2 / 0.0 ms  (average mu = 0.591, current mu = 0.467) allocation failure <--- JS stacktrace --->FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory 1: 0x10ce7f2e5 node::Buffer::New(v8::Isolate*, char*, unsigned long, void (*)(char*, void*), void*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 2: 0x10ce7f48a node::OnFatalError(char const*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 3: 0x10df76273 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 4: 0x10df76203 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 5: 0x10e041105 v8::internal::Heap::FinalizeIncrementalMarkingAtomically(v8::internal::GarbageCollectionReason) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 6: 0x10b19ccd1 v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 7: 0x10b191ddb v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 8: 0x10b1919e1 v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework] 9: 0x10b19160b v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]10: 0x10b1a1106 v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]11: 0x10cb5a4fd v8::internal::SetupIsolateDelegate::SetupHeap(v8::internal::Heap*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]12: 0x10cb5a923 v8::internal::SetupIsolateDelegate::SetupHeap(v8::internal::Heap*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]13: 0x10b19383d v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]14: 0x10b192dc3 v8::internal::MarkCompactCollector::EnsureSweepingCompleted() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]15: 0x10b184696 v8::internal::LocalHeap::PerformCollection() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]16: 0x10b1821f0 v8::internal::LocalHeap::PerformCollection() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]17: 0x10b163d3c v8::internal::Heap::MoveRange(v8::internal::HeapObject, v8::internal::CompressedObjectSlot, v8::internal::CompressedObjectSlot, int, v8::internal::WriteBarrierMode) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]18: 0x10b161129 v8::internal::Heap::CreateFillerObjectAt(unsigned long, int, v8::internal::ClearRecordedSlots) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]19: 0x10b15eb15 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]20: 0x10b169abb v8::internal::Heap::MaxRegularHeapObjectSize(v8::internal::AllocationType) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]21: 0x10b145970 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]22: 0x10b365349 v8::internal::RegExp::CompileForTesting(v8::internal::Isolate*, v8::internal::Zone*, v8::internal::RegExpCompileData*, v8::base::Flags<v8::internal::JSRegExp::Flag, int>, v8::internal::Handle<v8::internal::String>, v8::internal::Handle<v8::internal::String>, bool) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]23: 0x10b702398 v8::internal::SetupIsolateDelegate::SetupHeap(v8::internal::Heap*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]24: 0x2600198f3f 
log.ts:296   ERR Cannot call write after a stream was destroyed: Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at onwriteError (internal/streams/writable.js:389:22)
    at onwrite (internal/streams/writable.js:426:7)
    at internal/streams/destroy.js:60:7
    at Socket._destroy (net.js:678:5)
    at Socket.destroy (internal/streams/destroy.js:39:8)
    at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:95:12)
abstractExtensionService.ts:545 Extension host terminated unexpectedly. Code:  null  Signal:  SIGABRT
_onExtensionHostCrashed @ abstractExtensionService.ts:545
log.ts:296   ERR Extension host terminated unexpectedly. The following extensions were running: vscode.microsoft-authentication, vscode.extension-editing, vscode.markdown-language-features, yzhang.markdown-all-in-one, vscode.debug-auto-launch, vscode.git, kortina.vscode-markdown-notes, stkb.rewrap, foam.foam-vscode
notificationsAlerts.ts:40 Extension host terminated unexpectedly.
onDidChangeNotification @ notificationsAlerts.ts:40
log.ts:296   ERR Error: No authentication provider 'microsoft' is currently registered.
    at A.getSessions (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1946)
    at async vi.getAccounts (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:2189)
    at async vi.update (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:2189)
    at async vi.initialize (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:2189)
    at async vi.waitAndInitialize (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:2189)
log.ts:290  WARN Settings Sync: Accounts are not initialized
```
</details>

**Comment by jamie-brannan** [comment]
### Hello! I appear to have the same issue.

Yesterday, I did a bunch of disable / enable then uninstall/reboot/reinstall (then even a full machine reboot) but had the lower right corner GUI error message :

`An error occurred while bootstrapping Foam. TypeError: this.resourcesByName[name].push is not a function at FoamWorkspace.set (/Users/.../.vscode/extensions/foam.foam-vscode-0.13.2/node_modules/foam-core/dist/foam-core.cjs.development.js:2475:32) at /Users/.../.vscode/extensions/foam.foam-vscode-0.13.2/node_modules/foam-core/dist/foam-core.cjs.development.js:3058:31 at async Promise.all (index 270) at async Promise.all (index 0)`

I posted to the help channel of the Discord, but deleted it since soon after since it some hoe resolved

**TDLR** – _Today I have a clear cut 🔴 `Extension host terminated unexpectedly.` anytime the Foam extension is enabled. Although I'm not sure how to get the raw logs from the vsCode console, being on Big Sur 11.3 I imagine it's along the same lines._

**Comment by riccardoferretti** [comment]
do you have a reproducible set up for this problem? I can't seem to trigger it

**Comment by riccardoferretti** [comment]
As a reference, this occurs here: 
https://github.com/foambubble/foam/blob/master/packages/foam-core/src/model/workspace.ts#L61

Which is very puzzling as we are enforcing `this.resourcesByName[name]` to be an array just in the line above the error (and the only other place where `this.resourcesByName[name]` is set also seems correct).

Sharing as another pair of eyes might spot something I am missing.

**Comment by AndrewSheetMetal** [comment]
Just to provide a small reference for other flutter developers on this issue here...

I was not able to debug my flutter app, it hanged on the splash screen and showed "Assemble ..." in VS-Code with my CPU running hot.
After trying to downgrade flutter, dart and vs code itself, if still did not work.

After a while I detected a hint in the JS Console of VS-Code, that foam consumes much ressources.

I downgraded foam and everything works as before. 
It took me several hours to detect this side effect, because it was not quite obvious.
